vespa-cloud/cord-19-search/src/main/java/ai/vespa/example/cord19/searcher/RelatedPaperSearcherANN.java

```diff
@@ -49,46 +51,73 @@ public RelatedPaperSearcherANN(Linguistics linguistics) {
         }
-
-        Item queryTreeRoot = query.getModel().getQueryTree().getRoot();
-        String summary = query.getPresentation().getSummary();
-
-        query.getPresentation().setSummary(SUMMARY);
-        WordItem idFilter = new WordItem(id.toString(), "id", true);
-        query.getModel().getQueryTree().setRoot(idFilter);
-
-        Result result = execution.search(query);
-        execution.fill(result,SUMMARY);
-        Article article = extractFromResult(result);
-        if (article == null) {
-            return new Result(query);
+        if(traverseId != null)  {
+            id = traverseId;
         }
-
-        Query relatedQuery = generateRelatedQuery(article,includeAbstract,queryTreeRoot);
-        relatedQuery.getPresentation().setSummary(summary);
+        Article article = getArticle(id,execution,query);
+        Query relatedQuery = generateRelatedQuery(article,query,includeAbstract);
         relatedQuery.getPresentation().setBolding(false);
-        relatedQuery.setHits(query.getHits());
-        query.attachContext(relatedQuery);
+
         if(removeArticle) {
             NotItem notItem = new NotItem();
             notItem.addPositiveItem(relatedQuery.getModel().getQueryTree().getRoot());
-            notItem.addNegativeItem(idFilter);
+            notItem.addNegativeItem(new WordItem(id.toString(), "id", true));
             relatedQuery.getModel().getQueryTree().setRoot(notItem);
         }
+        relatedQuery.getRanking().setProfile(RELATED_RANKING_PROFILE);
         return execution.search(relatedQuery);
     }
 
+    private Article getArticle(Integer id, Execution execution,Query query) {
+        Query articleQuery = new Query();
+        query.attachContext(articleQuery);
+        articleQuery.getPresentation().setSummary(TENSOR_SUMMARY);
+        WordItem idFilter = new WordItem(id.toString(), "id", true);
+        articleQuery.getModel().getQueryTree().setRoot(idFilter);
+        articleQuery.getModel().setRestrict("doc");
+        articleQuery.setHits(1);
+        articleQuery.getRanking().setProfile("unranked");
+        Result articleResult = execution.search(articleQuery);
+        execution.fill(articleResult,TENSOR_SUMMARY);
+        return extractFromResult(articleResult);
+    }
+
+    /**
+     * Look for the related_to:x in the query tree
+     *
+     * @return
+     */
+    private Integer traverseQueryTree(Item item) {
+        if (item instanceof IntItem) {
+            IntItem word = (IntItem)item;
+            if(word.getIndexName().equals(RELATED_TO_FIELD)) {
+                return Integer.parseInt(word.getRawWord());
+            }
+        }
+        else if (item instanceof CompositeItem) {
+            CompositeItem compositeItem =(CompositeItem)item;
+            for(Item subItem: compositeItem.items()) {
+                Integer id = traverseQueryTree(subItem);
+                if(id != null) {
+                    compositeItem.removeItem(subItem);
```

Comment:
I don't think you are allowed to do this while looping over items? You loop with a listIterator instead and call remove on it.